core/commonMain/src/implementations/immutableList/PersistentVectorBuilder.kt

```diff
@@ -53,35 +53,57 @@ class PersistentVectorBuilder<E>(private var vector: PersistentList<E>,
         return rootSize(size)
     }
 
+    private fun tailSize(size: Int): Int {
+        if (size <= MAX_BUFFER_SIZE) {
+            return size
+        }
+        return size - rootSize(size)
+    }
+
+    private fun tailSize(): Int {
+        return tailSize(size)
+    }
+
+    private fun isMutable(buffer: Array<Any?>): Boolean {
+        return buffer.size == MUTABLE_BUFFER_SIZE && buffer[MUTABLE_BUFFER_SIZE - 1] === ownership
+    }
+
     /**
      * Checks if [buffer] is mutable and returns it or its mutable copy.
      */
     private fun makeMutable(buffer: Array<Any?>?): Array<Any?> {
         if (buffer == null) {
-            val newBuffer = arrayOfNulls<Any?>(MUTABLE_BUFFER_SIZE)
-            newBuffer[MUTABLE_BUFFER_SIZE - 1] = ownership
-            return newBuffer
+            return mutableBuffer()
         }
-        if (buffer.size != MUTABLE_BUFFER_SIZE || buffer[MUTABLE_BUFFER_SIZE - 1] !== ownership) {
-            val newBuffer = arrayOfNulls<Any?>(MUTABLE_BUFFER_SIZE)
-            buffer.copyInto(newBuffer, endIndex = buffer.size.coerceAtMost(MAX_BUFFER_SIZE))
-            newBuffer[MUTABLE_BUFFER_SIZE - 1] = ownership
-            return newBuffer
+        if (isMutable(buffer)) {
+            return buffer
         }
-        return buffer
+        return buffer.copyInto(mutableBuffer(), endIndex = buffer.size.coerceAtMost(MAX_BUFFER_SIZE))
+    }
+
+    private fun makeMutableShiftingRight(buffer: Array<Any?>, distance: Int): Array<Any?> {
+        if (isMutable(buffer)) {
+            return buffer.copyInto(buffer, distance, 0, MAX_BUFFER_SIZE - distance)
+        }
+        return buffer.copyInto(mutableBuffer(), distance, 0, MAX_BUFFER_SIZE - distance)
     }
 
     private fun mutableBufferWith(element: Any?): Array<Any?> {
-        val buffer = arrayOfNulls<Any?>(MUTABLE_BUFFER_SIZE)
+        val buffer = mutableBuffer()
```

Comment:
I think we could tolerate a little less code reuse percentage and inline `mutableBuffer()` here ;)

Reply:
ok